### PR TITLE
Implement the "-Werror" compiler flag handling when emitting compiler diagnostics as problems

### PR DESCRIPTION
**Pull request by hegyibalint**
Closes [#27948](https://github.com/gradle/gradle/issues/27948) by solving the transforming/filtering of diagnostics.

### CODE REVIEW

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 1]
```diff
@@ -35,20 +40,47 @@ public class DiagnosticToProblemListener implements DiagnosticListener<JavaFileO
...
+
+    public void addSeverityOverride(Diagnostic.Kind fromKind, Diagnostic.Kind toKind) {
+        severityOverrides.put(fromKind, toKind);
+    }
+
     @Override
     public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
+        // If any of the filters return true, we should not report the diagnostic
+        if (diagnosticFilters.stream().anyMatch(filter -> filter.test(diagnostic))) {
+            return;
```
Comment (tresat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:70]: 💅 Should this `report` method be renamed `onProblem` or `maybeReport` or something, as reporting now depends on a diagnostic surviving the filters?

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 2]
```diff
@@ -35,20 +40,47 @@ public class DiagnosticToProblemListener implements DiagnosticListener<JavaFileO
...
     private final ProblemReporter problemReporter;
 
+    /** A collection of filters that will serve as a predicate to determine if a diagnostic should be reported or not.*/
+    private final Collection<Predicate<Diagnostic<? extends JavaFileObject>>> diagnosticFilters = new ArrayList<>();
+
+    /**
+     * A map of explicit overrides between {@link Diagnostic.Kind} and {@link Severity}.
+     * <p>
+     * This will take precedence over the default mapping defined in {@link #mapKindToSeverity(Diagnostic.Kind)}.
+     */
```
Comment (tresat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:50]: 💭 These javadocs would probably be better on the public methods, than hidden internally on the fields.

**Review by tresat — APPROVED** [review]

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 2 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:50]: You have a point, I will put them there

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 1 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:70]: Meanwhile your reasoning is sound, I cannot as this method comes from [`DiagnosticListener`](https://github.com/openjdk/jdk/blob/master/src/java.compiler/share/classes/javax/tools/DiagnosticListener.java)

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 2 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java]: Moved!

**Review by donat — APPROVED** [review]

**Comment by hegyibalint** [comment]
@bot-gradle merge

**Comment by donat** [comment]
Let's not do this, at least not without broader discussion on the topic.

**Pull request closed without merging**